Livekit

mathisbarthere opened the pull request:
## Purpose

This PR aims to replace Jitsi with LiveKit in the Magnify project.

Commit:
LiveKit token generation

Commit:
Add Livekit env variables

raphfaur committed:
:green_heart:(LiveKit) Implement Room and PreJoin

Commit:
Merge branch 'livekit-token' into livekit

Fusion of token and call review

raphfaur committed:
:sparkles:(frontend) Navigate home on disconnect

raphfaur committed:
:hammer:(backend) Allow publishing tracks

raphfaur committed:
:wrench:(build) Fix build

raphfaur committed:
(build) Lock file update

raphfaur committed:
:construction_worker:(ci/cd) Setup production build ci

raphfaur committed:
Update build registry

raphfaur committed:
CI test

raphfaur committed:
Change context

raphfaur committed:
Delete context

raphfaur committed:
Custom registry

raphfaur committed:
Merge pull request #1 from numerique-gouv/github-action

Github action setup

raphfaur committed:
Pass user choices as a context

raphfaur committed:
:wrench: Fix build

raphfaur committed:
:white_check_mark: Implement test for RoomLivekitView

raphfaur committed:
Remove log

Co-authored-by: Sacha Dupuydauby <gh.4eczg@slmail.me>

raphfaur committed:
Remove log

Co-authored-by: Sacha Dupuydauby <gh.4eczg@slmail.me>

sleiderr committed:
removed hardcoded message

sleiderr committed:
only execute workflow when merging into livekit

raphfaur committed:
Implement tests for LiveKitView

raphfaur committed:
Merge pull request #2 from numerique-gouv/livekit-context-provider

Provide user choices in context

raphfaur committed:
Merge pull request #3 from numerique-gouv/action-pr

:construction_worker: Only run workflow when merging into a specific branch (LiveKit)

sleiderr committed:
Merge pull request #4 from numerique-gouv/livekit-test

Livekit test

Commit:
change server url

raphfaur committed:
Set livekit domain as an env variable

raphfaur committed:
Merge pull request #5 from numerique-gouv/livekit-env-settings

Set livekit domain as an env variable

raphfaur committed:
(backend) Allow guest users to set username

raphfaur committed:
(frontend) Allow guest users to set username during prejoin

raphfaur committed:
Merge pull request #7 from numerique-gouv/(fix)livekit-guest-join

:white_check_mark: Allow guests on public rooms closes #6

raphfaur committed:
New action to approve pull request

raphfaur committed:
Add job name

raphfaur committed:
Livekit target branch

raphfaur committed:
Remove if condition

raphfaur committed:
Fix build

raphfaur committed:
Merge pull request #8 from numerique-gouv/(actions)livekit-merge-validation

Create a job to block merge in case it would fail

raphfaur committed:
:rocket: (feat) Implement admin role

raphfaur committed:
Merge branch 'livekit' into livekit-admin-control

sleiderr committed:
Set image tag based on source branch name (#9)

raphfaur committed:
(refactor) Set API url variable as a window config variable

raphfaur committed:
(fix) Joining a room will now set the right video and audio state

raphfaur committed:
(feat) Implement screen sharing mute and admin-only options

raphfaur committed:
(feat) Remove participant and responsive layout

raphfaur committed:
`ControlBar` controls depending on allowed sources

raphfaur committed:
(frontend) Change screen sharing icon

raphfaur committed:
(feat) Replace default control bar with custom `ControlBar` including `ParticipantLayoutToggle`

raphfaur committed:
User removal confirm and adaptative `MagnifyControlBar`

raphfaur committed:
:sparkles: Improve UI

raphfaur committed:
(fix) ref for `PopOver`

Commit:
(feat) add WaitingRoom and RaisingHand

raphfaur committed:
(feat) Implements an `EventHandler` context in order to provide notifications utilities

raphfaur committed:
(feat) Implement join and mute notification

raphfaur committed:
Merge branch 'livekit-admin-control' into livekit-waiting-room

raphfaur committed:
(fix) Join and build

raphfaur committed:
(action) Workflow dispatch

raphfaur committed:
(fix) Hidden overflow

raphfaur committed:
Button are now disabled instead of being deleted

raphfaur committed:
Delete log statements

raphfaur committed:
(feat) Implement layout selections

raphfaur committed:
(fix) Hide `AdminBulkActions`component for non-admin users

sleiderr committed:
feat(waiting-room): only display waiting room when enabled in magnify

raphfaur committed:
Responsive layout

raphfaur committed:
Swipe fix

raphfaur committed:
(fix) Prejoin choices are now applied

raphfaur committed:
Focus indicator

raphfaur committed:
(mobile) Improve mobile UX

raphfaur committed:
Remove unused import

raphfaur committed:
Merge pull request #13 from numerique-gouv/livekit-mobile-ui

Improve mobile UX

Commit:
(feat): add color for raised hand, fix layouts and bulk actions

Commit:
fix code not supposed to exist

sleiderr committed:
fix(waiting-room): fix waiting room handler for newly created rooms

sleiderr committed:
fix(waiting-room): removed debug print

raphfaur committed:
Merge pull request #12 from numerique-gouv/livekit-autojoin-no-waiting-room

feat(waiting-room): only display waiting room when enabled in magnify

Commit:
add tick for layout

sleiderr committed:
feat(themes): theme update (killer features)

Commit:
add tick for layout

sleiderr committed:
feat(themes): fixed control bar buttons color

sleiderr committed:
fix(themes): fix invalid theme colors

sleiderr committed:
fix(ui): participant layout size

raphfaur committed:
(fix) Scroll participant layout

raphfaur committed:
(fix) Joining temp room

raphfaur committed:
Merge pull request #11 from numerique-gouv/livekit-waiting-room

LiveKit full implementation

Commit:
minors fixes, including translation in french

raphfaur committed:
(fix) Fix mobile UI

raphfaur committed:
Disable pin of inactive tracks

raphfaur committed:
Merge pull request #14 from numerique-gouv/livekit-waiting-room

Minors fixes, including french translation

raphfaur committed:
(fix) Addmin controls are now visible on mobile view

raphfaur committed:
(fix) Admin controls are now visible on mobile view

raphfaur committed:
Merge pull request #15 from numerique-gouv/livekit-waiting-room

Typo

raphfaur committed:
New landing page title e2e encryption

raphfaur committed:
(github-actions) Build on push

raphfaur committed:
(github-actions) Specify branche

raphfaur committed:
(github-actions) Remove if statement

raphfaur committed:
(fix) Typo

raphfaur committed:
Change nading page title

raphfaur committed:
(feat) Implement magnify configuration in LiveKit rooms

raphfaur committed:
(fix) Type changes

raphfaur committed:
Merge pull request #17 from numerique-gouv/livekit-settings-support

Implement magnify configuration in LiveKit rooms

raphfaur committed:
(refactor) Change project structure

raphfaur committed:
(fix) Circular import and prejoin fix

Commit:
add encryption

Commit:
fix LockIcon

raphfaur committed:
(refactor) Remove non necessary inline css and rename files

raphfaur committed:
Merge pull request #18 from numerique-gouv/livekit-refactor

(refactor) Change project structure

Commit:
add encryption

Commit:
fix type for build

raphfaur committed:
(doc) Update documentation and clean remaining jitsi occurence

raphfaur committed:
Merge pull request #19 from numerique-gouv/livekit-encryption

Livekit encryption

raphfaur committed:
Merge pull request #20 from numerique-gouv/livekit-readme

Livekit Cleanup

Commit:
improve encryption in control baré

raphfaur committed:
(fix) Various fixes

raphfaur committed:
Merge pull request #22 from numerique-gouv/livekit-encryption

Improve `ControlBar`

raphfaur committed:
Merge pull request #21 from numerique-gouv/livekit-fixes

Various fixes

raphfaur committed:
Admin controls and frontend fixes

raphfaur committed:
Merge pull request #23 from numerique-gouv/livekit-fixes

Admin controls and frontend fixes

raphfaur committed:
(fix) Layout default display

raphfaur committed:
Merge pull request #24 from numerique-gouv/livekit-fixes

(fix) Layout default display

Commit:
add part of translation

Commit:
add translation

gamma3591 committed:
Merge branch 'livekit' into livekit-encryption

gamma3591 committed:
fix typo

gamma3591 committed:
remove comments

gamma3591 committed:
Merge pull request #25 from numerique-gouv/livekit-encryption

Livekit translation

manuhabitela committed:
(chore) docker-compose file: `version` is obsolete

docker compose doesn't check the listed `version` anymore, remove it to
prevent the CLI to warn us on every command

manuhabitela committed:
(chore) include a testing livekit server in docker env

one thing not mentioned a lot in the current docs is that magnify
requires a LiveKit server to talk to in order to work.

No livekit server was included so it required each developer to
understand this on their own.

Now we include a testing livekit server in the docker containers. This
makes running the app from scratch working even if we are not aware of
the internals.

the common env vars changed to match the default livekit server
settings.

manuhabitela committed:
(chore) new `make dev` command to ease up local dev:

`make dev` is meant to replace `make run` locally, it starts everything
needed to develop locally. Contrary to `make run` it doesn't start the
demo "production" front-end, but it starts the vite live reload server

manuhabitela committed:
(chore) `make boostrap`: do not start the app

personal opinion: it feels weird that the bootstrap command runs the
app. Removing this behavior so that we can then decide to run the app in
dev mode or not (via `make run` or `make dev`) after running `make
bootstrap`. This makes `make boostrap` better fit the intent from my
pov: a command you run after updating your repo to install
deps/migrations etc.

manuhabitela commented:
:wave: 

Thanks a lot for all this, it's a lot of great work :)

Allowed myself to add a few things to ease-up local development. Now a livekit testing server is included as a docker service.

@lebaudantoine you should be able to run a local instance by blinding following the readme now.

Commit:
(chore) stop forcing platform for docker images

Forcing `platform: linux/amd64` for the docker image causes compatibility
issues and performance degradation on Mac ARM chips (M1/M2). Removing the
platform specification allows Docker to select the appropriate architecture
automatically, ensuring better performance and compatibility.

lebaudantoine commented:
Great work 🤟🙏
I have pushed a commit to make it compatible with my platform.

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
     """Get JWT tokens for user authentication."""
     refresh = RefreshToken.for_user(user)
-
+    test = settings.LIVEKIT_CONFIGURATION["livekit_token_expiration_seconds"]
```
useless variable

lebaudantoine commented on src/frontend/sandbox/public/assets/logo-tu.svg:
could you let me know what `tu` stands for?

lebaudantoine commented on src/frontend/packages/core/vite.config.ts:
```diff
@@ -30,6 +30,7 @@ export default defineConfig({
         },
       },
     },
+    
```
Let's try to avoid creating "noisy diffs" – these are changes that don't add any real value to the code. By keeping our diffs clean and focused, we make it easier for everyone to review and understand the meaningful changes.

lebaudantoine commented on src/frontend/packages/core/src/views/rooms/livekit/index.test.tsx:
```diff
@@ -0,0 +1,83 @@
+        const internalCamera = screen.getAllByText("Internal camera")
+        assert(internalCamera.length == 2)
+    })
+})
```
EOF, few files are missing a proper EOF.

lebaudantoine commented on src/frontend/sandbox/src/cunningham-tokens.ts:
```diff
@@ -1,101 +1 @@
-    },
-  },
-};
+export const tokens = {"theme":{"colors":{"primary-text":"#FFFFFF","primary-100":"#EBF2FC","primary-200":"#8CB5EA","primary-300":"#5894E1","primary-400":"#114390","primary-500":"#114390","primary-600":"#0556BF","primary-700":"#044395","primary-800":"#033474","primary-900":"#022858","secondary-text":"#555F6B","secondary-100":"#F2F7FC","secondary-200":"#EBF3FA","secondary-300":"#E2EEF8","secondary- ...
```
formating issue?

lebaudantoine commented on src/magnify/apps/core/api/__init__.py:
```diff
@@ -48,7 +48,7 @@ def exception_handler(exc, context):
     """Returns the frontend configuration dict as configured in settings."""
     frontend_configuration = {
-        "JITSI_DOMAIN": settings.JITSI_CONFIGURATION["jitsi_domain"],
+        "LIVEKIT_DOMAIN": settings.LIVEKIT_CONFIGURATION["livekit_domain"],
```
I understand it's legacy, but it does feel a bit odd to prefix the domain with livekit for configurations that are already in a settings section also prefixed with livekit. Just a thought! Feel free to disregard my comment.

lebaudantoine commented on src/frontend/sandbox/package.json:
It's a good practice to pin dependencies! This helps ensure that our project remains stable by avoiding unexpected changes from updated versions, which can sometimes introduce bugs or compatibility issues.

lebaudantoine commented on src/frontend/packages/core/src/types/api/room/index.ts:
```diff
@@ -1,11 +1,11 @@
 
-export interface RoomResponse extends Room {}
+export interface RoomResponse extends Room { }
 export interface RoomsResponse {
   count: number;
   results: Room[];
 }
 export interface CreateRoomData {
   name: string;
 }
-export interface UpdateRoomData extends Room {}
+export interface UpdateRoomData extends Room { }
```
were these diffs made by a linter?

lebaudantoine commented on src/frontend/packages/core/src/cunningham-tokens.css:
```diff
@@ -4,7 +4,7 @@
   --c--theme--colors--primary-300: #5894e1;
   --c--theme--colors--primary-400: #377fdb;
-  --c--theme--colors--primary-500: #055fd2;
+  --c--theme--colors--primary-500: #000000;
```
Let's make sure the 500 variation stays lighter than the higher ones.

lebaudantoine commented on src/frontend/packages/core/src/components/livekit/display/pannel/ParticipantsLayout/style.css:
Overall, using PascalCase for CSS class naming is not a common practice, but that's fine

lebaudantoine commented on src/magnify/apps/core/models.py:
```diff
@@ -516,7 +525,7 @@ def __str__(self):
     @property
     def jitsi_name(self):
         """The name used as Jitsi room for this meeting."""
```
I think this method naming and documentation are outdated. Also, tests still mentioned `jitsi`

lebaudantoine commented on src/magnify/apps/core/serializers/rooms.py:
```diff
@@ -1,4 +1,6 @@
 """Serializers for the core Magnify app."""
+import random
+import string
```
I think these imports are useless

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
+    token_payload = api.AccessToken(
+        settings.LIVEKIT_CONFIGURATION["livekit_api_key"],
+        settings.LIVEKIT_CONFIGURATION["livekit_api_secret"],
+    ).with_identity(identity).with_name(username).with_grants(video_grants).with_ttl(timedelta(seconds=expiration_seconds)).with_metadata(f"{{ \"admin\" : {json.dumps(is_admin)}}}")
```
Sources might need to be linted.

lebaudantoine replied:
Why do we explicitly need to set `with_ttl`

lebaudantoine commented on yarn.lock:
This file might be useless

lebaudantoine reviewed (commented):
Hey, sorry for the nit-picking

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
+    grants = api.VideoGrants(room_join=True, room=room, can_publish=False, can_subscribe=False, room_admin=is_admin,
+                             can_update_own_metadata=True, can_publish_sources=["camera", "microphone", "screen_share", "screen_share_audio"])
+
+    if is_temp_room:
+        grants = api.VideoGrants(room_join=True, room=room, can_publish=True, can_subscribe=True, room_admin=is_admin,
+                                 can_update_own_metadata=True, can_publish_sources=["camera", "microphone", "screen_share", "screen_share_audio"])
+        return grants
```
you can move this early return be first

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
+    return sources
+
+
+def create_video_grants(room: string, is_admin=False, is_temp_room=True):
```
We can simplify these calls to the VideoGrants constructor by creating a base list of arguments. Then, depending on your conditions, you can enrich or override this base list with additional arguments.

lebaudantoine commented on src/magnify/apps/core/models.py:
```diff
@@ -372,10 +372,19 @@ class Room(Resource):
-        default=dict,
-        verbose_name=_("Jitsi room configuration"),
+        default=lambda : {
+            "roomPassword": "",
+            "askForPassword": False,
+            "enableLobbyChat": True,
+            "waitingRoomEnabled": True,
+            "startWithAudioMuted": False,
+            "startWithVideoMuted": True,
+            "askForAuthentication": True,
+            "screenSharingEnabled": True
```
wip: understand if we explicitly need to pass all of these configurations.

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
-    }
+    video_grants = create_video_grants(room, is_admin, is_temp_room)
+    token_payload = api.AccessToken(
+        settings.LIVEKIT_CONFIGURATION["livekit_api_key"],
+        settings.LIVEKIT_CONFIGURATION["livekit_api_secret"],
```
note: should automatically use the right env vars, according to [the doc](https://github.com/livekit/python-sdks?tab=readme-ov-file#generating-an-access-token)

If wrapped by django, we need to explicitly set them

lebaudantoine commented on src/magnify/apps/core/utils.py:
```diff
@@ -44,49 +50,70 @@ def get_nth_week_number(original_date):
+        return grants
+
+    try:
+        roomData = models.Room.objects.get(id=uuid.UUID(room))
```
snakeCase

lebaudantoine commented on src/frontend/packages/core/src/components/AppRouter/index.tsx:
```diff
@@ -36,11 +36,11 @@ export const AppRouter = () => {
           index: true,
           element: (
-            <SimpleLayout urlLogo="/assets/logo-fun-mooc.svg">
+            <SimpleLayout urlLogo="/assets/logo-fun-mooc.svg" >
```
outdated
